File: services/control-plane/internal/vaultclient/vaultclient.go
// Package vaultclient implements Vault's Kubernetes auth method login
// flow and Transit sign/verify calls — the mechanism registry/ uses to
// sign manifests at RegisterModelManifest (Step E) and verify them at
// GetModelManifest, and that Step J's IssueProviderToken will reuse for
// its own Vault access.
//
// Kubernetes auth, not a static Vault token in a Secret: control-plane
// authenticates to Vault using its own pod's projected ServiceAccount
// token (the cluster's own identity for it), which Vault validates via
// the Kubernetes TokenReview API against the bound role configured in
// scripts/vault-setup-control-plane.sh. This is the user's own explicit
// requirement — reusing a static credential to reach Vault would just
// move the exact problem Vault is replacing (a stored secret) one layer
// down.
package vaultclient

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"sync"
	"time"
)

const defaultSATokenPath = "/var/run/secrets/kubernetes.io/serviceaccount/token"

// Signer is the interface registry/ depends on — Client below is the real
// Vault-Transit-backed implementation; tests use a fake.
type Signer interface {
	Sign(ctx context.Context, keyName string, input []byte) (signature string, err error)
	Verify(ctx context.Context, keyName string, input []byte, signature string) (valid bool, err error)
}

// Client logs in via Vault's Kubernetes auth method on first use and
// re-logs-in once the cached token is within its own renewal buffer of
// expiring — off the hot path (Phase-04.txt's own framing), so a simple
// lazy-refresh-before-call is proportionate; no background renewal loop.
type Client struct {
	addr         string
	k8sRole      string
	saTokenPath  string
	httpClient   *http.Client
	mu           sync.Mutex
	vaultToken   string
	tokenExpires time.Time
}

func New(addr, k8sRole string) *Client {
	return &Client{
		addr:        addr,
		k8sRole:     k8sRole,
		saTokenPath: defaultSATokenPath,
		httpClient:  &http.Client{Timeout: 10 * time.Second},
	}
}

type k8sLoginRequest struct {
	Role string `json:"role"`
	JWT  string `json:"jwt"`
}

type vaultAuthResponse struct {
	Auth struct {
		ClientToken   string `json:"client_token"`
		LeaseDuration int    `json:"lease_duration"`
	} `json:"auth"`
	Errors []string `json:"errors"`
}

// login performs the Kubernetes auth method's login flow: read this pod's
// own projected ServiceAccount token (the cluster-vouched identity, not a
// stored credential) and exchange it for a Vault token scoped to
// k8sRole's policy.
func (c *Client) login(ctx context.Context) error {
	jwt, err := os.ReadFile(c.saTokenPath)
	if err != nil {
		return fmt.Errorf("reading service account token: %w", err)
	}

	body, err := json.Marshal(k8sLoginRequest{Role: c.k8sRole, JWT: string(jwt)})
	if err != nil {
		return fmt.Errorf("marshaling login request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.addr+"/v1/auth/kubernetes/login", bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("building login request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("calling Vault kubernetes login: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("reading login response: %w", err)
	}

	var authResp vaultAuthResponse
	if err := json.Unmarshal(respBody, &authResp); err != nil {
		return fmt.Errorf("decoding login response (status %d): %w", resp.StatusCode, err)
	}
	if resp.StatusCode != http.StatusOK || authResp.Auth.ClientToken == "" {
		return fmt.Errorf("vault kubernetes login failed (status %d): %v", resp.StatusCode, authResp.Errors)
	}

	c.vaultToken = authResp.Auth.ClientToken
	// 30s renewal buffer: re-login proactively rather than have a signing
	// call fail on an edge-of-expiry token.
	c.tokenExpires = time.Now().Add(time.Duration(authResp.Auth.LeaseDuration)*time.Second - 30*time.Second)
	return nil
}

func (c *Client) ensureToken(ctx context.Context) (string, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.vaultToken != "" && time.Now().Before(c.tokenExpires) {
		return c.vaultToken, nil
	}
	if err := c.login(ctx); err != nil {
		return "", err
	}
	return c.vaultToken, nil
}

type transitSignRequest struct {
	Input string `json:"input"`
}

type transitSignResponse struct {
	Data struct {
		Signature string `json:"signature"`
	} `json:"data"`
	Errors []string `json:"errors"`
}

// Sign returns a Transit signature over input, in Vault's own
// "vault:v<version>:<base64>" format — opaque to callers, passed back to
// Verify unmodified.
func (c *Client) Sign(ctx context.Context, keyName string, input []byte) (string, error) {
	token, err := c.ensureToken(ctx)
	if err != nil {
		return "", fmt.Errorf("vault auth: %w", err)
	}

	reqBody, err := json.Marshal(transitSignRequest{Input: base64.StdEncoding.EncodeToString(input)})
	if err != nil {
		return "", fmt.Errorf("marshaling sign request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.addr+"/v1/transit/sign/"+keyName, bytes.NewReader(reqBody))
	if err != nil {
		return "", fmt.Errorf("building sign request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Vault-Token", token)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("calling transit sign: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("reading sign response: %w", err)
	}

	var signResp transitSignResponse
	if err := json.Unmarshal(respBody, &signResp); err != nil {
		return "", fmt.Errorf("decoding sign response (status %d): %w", resp.StatusCode, err)
	}
	if resp.StatusCode != http.StatusOK || signResp.Data.Signature == "" {
		return "", fmt.Errorf("transit sign failed (status %d): %v", resp.StatusCode, signResp.Errors)
	}
	return signResp.Data.Signature, nil
}

type transitVerifyRequest struct {
	Input     string `json:"input"`
	Signature string `json:"signature"`
}

type transitVerifyResponse struct {
	Data struct {
		Valid bool `json:"valid"`
	} `json:"data"`
	Errors []string `json:"errors"`
}

// Verify checks signature against input under keyName. A malformed or
// tampered signature, or a signature over different input, returns
// (false, nil) — a definite negative answer, not an error — matching
// Vault's own transit/verify semantics (a well-formed request that
// legitimately doesn't verify is not a call failure). A genuine call
// failure (network, auth, bad key name) returns a non-nil error.
func (c *Client) Verify(ctx context.Context, keyName string, input []byte, signature string) (bool, error) {
	token, err := c.ensureToken(ctx)
	if err != nil {
		return false, fmt.Errorf("vault auth: %w", err)
	}

	reqBody, err := json.Marshal(transitVerifyRequest{
		Input:     base64.StdEncoding.EncodeToString(input),
		Signature: signature,
	})
	if err != nil {
		return false, fmt.Errorf("marshaling verify request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.addr+"/v1/transit/verify/"+keyName, bytes.NewReader(reqBody))
	if err != nil {
		return false, fmt.Errorf("building verify request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-Vault-Token", token)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return false, fmt.Errorf("calling transit verify: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return false, fmt.Errorf("reading verify response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		var errResp transitVerifyResponse
		_ = json.Unmarshal(respBody, &errResp)
		// A malformed/garbage signature (e.g. an unsigned manifest whose
		// signature column is empty or nonsense) is Vault's own
		// invalid-signature 400, not a call failure — treated as a
		// definite "no", same as a well-formed-but-wrong signature.
		if resp.StatusCode == http.StatusBadRequest {
			return false, nil
		}
		return false, fmt.Errorf("transit verify failed (status %d): %v", resp.StatusCode, errResp.Errors)
	}

	var verifyResp transitVerifyResponse
	if err := json.Unmarshal(respBody, &verifyResp); err != nil {
		return false, fmt.Errorf("decoding verify response: %w", err)
	}
	return verifyResp.Data.Valid, nil
}

type kvReadResponse struct {
	Data struct {
		Data map[string]any `json:"data"`
	} `json:"data"`
	Errors []string `json:"errors"`
}

// ReadProviderSecret reads secret/provider/{provider} (KV v2 — Step I's
// own scripts/vault-load-provider-keys.sh writes to this exact path) and
// returns its api_key field. Used by Step J's IssueProviderToken — never
// logs the returned value, only that a read happened.
func (c *Client) ReadProviderSecret(ctx context.Context, provider string) (string, error) {
	token, err := c.ensureToken(ctx)
	if err != nil {
		return "", fmt.Errorf("vault auth: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, c.addr+"/v1/secret/data/provider/"+provider, nil)
	if err != nil {
		return "", fmt.Errorf("building secret read request: %w", err)
	}
	req.Header.Set("X-Vault-Token", token)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("calling secret read: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("reading secret read response: %w", err)
	}

	if resp.StatusCode == http.StatusNotFound {
		return "", fmt.Errorf("no secret stored at secret/provider/%s", provider)
	}
	if resp.StatusCode != http.StatusOK {
		var errResp kvReadResponse
		_ = json.Unmarshal(respBody, &errResp)
		return "", fmt.Errorf("secret read failed (status %d): %v", resp.StatusCode, errResp.Errors)
	}

	var kvResp kvReadResponse
	if err := json.Unmarshal(respBody, &kvResp); err != nil {
		return "", fmt.Errorf("decoding secret read response: %w", err)
	}

	apiKey, ok := kvResp.Data.Data["api_key"].(string)
	if !ok || apiKey == "" {
		return "", fmt.Errorf("secret/provider/%s has no non-empty api_key field", provider)
	}
	return apiKey, nil
}

File: services/edge-gateway/src/vault_client.rs
//! vault_client — Phase-04 Step R: edge-gateway's OWN Vault access,
//! entirely independent of control-plane's (Go) and data-plane's
//! (Python) — same reasoning as data-plane's own vault_client module doc:
//! model_registry's cache must verify a manifest's signature ITSELF
//! before trusting it, never because "it came from etcd" or "another
//! service already checked it." A compromised etcd or a bad publish must
//! not be able to make edge-gateway route on an invalid manifest.
//!
//! Kubernetes auth, not a static Vault token: edge-gateway authenticates
//! using its own pod's projected ServiceAccount token, validated via
//! Vault's Kubernetes auth method against the bound role
//! scripts/vault-setup-edge-gateway.sh configures.
//!
//! VERIFY ONLY: edge-gateway's Vault policy (edge-gateway-manifest-verify)
//! grants transit/verify on model-manifest-signing but NOT transit/sign —
//! same asymmetric capability data-plane's own policy already established
//! (Step Q): even a fully compromised edge-gateway pod cannot forge a new
//! signed manifest, only check existing signatures.

use std::time::{Duration, Instant};

use base64::Engine;
use base64::engine::general_purpose::STANDARD as BASE64;
use serde_json::json;
use tokio::sync::Mutex;

const SA_TOKEN_PATH: &str = "/var/run/secrets/kubernetes.io/serviceaccount/token";

/// Renewal buffer mirrors data-plane's own vault_client.Client — re-login
/// before the cached token's TTL actually expires, not exactly at it.
const RENEWAL_BUFFER: Duration = Duration::from_secs(30);

#[derive(Debug, Clone)]
pub enum VaultError {
    /// A genuine call failure (network, auth, bad key name, token file
    /// unreadable) — distinct from a well-formed verify call that
    /// legitimately answers false.
    Call(String),
}

impl std::fmt::Display for VaultError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            VaultError::Call(msg) => write!(f, "{msg}"),
        }
    }
}

struct TokenState {
    token: String,
    expires_at: Instant,
}

pub struct VaultClient {
    addr: String,
    k8s_role: String,
    http: reqwest::Client,
    token: Mutex<Option<TokenState>>,
}

impl VaultClient {
    pub fn new(addr: String, k8s_role: String, http: reqwest::Client) -> Self {
        Self { addr, k8s_role, http, token: Mutex::new(None) }
    }

    async fn login(&self) -> Result<TokenState, VaultError> {
        let jwt = tokio::fs::read_to_string(SA_TOKEN_PATH)
            .await
            .map_err(|e| VaultError::Call(format!("reading service account token: {e}")))?;

        let resp = self
            .http
            .post(format!("{}/v1/auth/kubernetes/login", self.addr))
            .json(&json!({"role": self.k8s_role, "jwt": jwt.trim()}))
            .send()
            .await
            .map_err(|e| VaultError::Call(format!("calling vault kubernetes login: {e}")))?;

        let status = resp.status();
        let body: serde_json::Value = resp
            .json()
            .await
            .map_err(|e| VaultError::Call(format!("decoding login response: {e}")))?;

        let client_token = body
            .get("auth")
            .and_then(|a| a.get("client_token"))
            .and_then(|t| t.as_str());

        match client_token {
            Some(token) if status.is_success() => {
                let lease_secs = body["auth"]["lease_duration"].as_u64().unwrap_or(0);
                let ttl = Duration::from_secs(lease_secs).saturating_sub(RENEWAL_BUFFER);
                Ok(TokenState { token: token.to_string(), expires_at: Instant::now() + ttl })
            }
            _ => Err(VaultError::Call(format!(
                "vault kubernetes login failed (status {status}): {:?}",
                body.get("errors")
            ))),
        }
    }

    async fn ensure_token(&self) -> Result<String, VaultError> {
        let mut guard = self.token.lock().await;
        if let Some(state) = guard.as_ref() {
            if Instant::now() < state.expires_at {
                return Ok(state.token.clone());
            }
        }
        let state = self.login().await?;
        let token = state.token.clone();
        *guard = Some(state);
        Ok(token)
    }

    /// Verifies signature over payload under key_name via Vault Transit.
    /// Returns Ok(false) for a well-formed request that legitimately
    /// doesn't verify (tampered/unsigned/garbage signature) — a definite
    /// negative answer, not a call failure. Returns Err only for a
    /// genuine call failure.
    pub async fn verify(
        &self,
        key_name: &str,
        payload: &[u8],
        signature: &str,
    ) -> Result<bool, VaultError> {
        let token = self.ensure_token().await?;

        let resp = self
            .http
            .post(format!("{}/v1/transit/verify/{key_name}", self.addr))
            .header("X-Vault-Token", token)
            .json(&json!({"input": BASE64.encode(payload), "signature": signature}))
            .send()
            .await
            .map_err(|e| VaultError::Call(format!("calling transit verify: {e}")))?;

        let status = resp.status();
        if status.as_u16() == 400 {
            // Vault's own invalid-signature response — same "well-formed
            // request, definite no" treatment the Go and Python
            // vault_client implementations both give this exact status.
            return Ok(false);
        }

        let body: serde_json::Value = resp
            .json()
            .await
            .map_err(|e| VaultError::Call(format!("decoding verify response: {e}")))?;

        if !status.is_success() {
            return Err(VaultError::Call(format!(
                "transit verify failed (status {status}): {:?}",
                body.get("errors")
            )));
        }

        Ok(body["data"]["valid"].as_bool().unwrap_or(false))
    }
}
